cpp/Contact.hpp

```diff
@@ -1350,6 +1316,119 @@ class Contact
     return {std::move(c), cstride};
   }
 
+  /// Compute inward surface normal at Pi(x)
+  /// @param[in] orgin_meshtag - surface on which to integrate
+  /// @param[out] c - normals ny packed on facets.
+  std::pair<std::vector<PetscScalar>, int>
+  pack_ny(int origin_meshtag, const xtl::span<const PetscScalar> gap)
+  {
+
+    // Mesh info
+    auto mesh = _marker->mesh();             // mesh
+    const int gdim = mesh->geometry().dim(); // geometrical dimension
+    const int tdim = mesh->topology().dim();
+    const int fdim = tdim - 1;
+    auto cmap = mesh->geometry().cmap();
+    auto x_dofmap = mesh->geometry().dofmap();
+    const xt::xtensor<double, 2>& mesh_geometry = mesh->geometry().x();
+    xt::xtensor<std::int32_t, 3>* map;
+    std::vector<xt::xtensor<double, 2>>* q_phys_pt;
+    xt::xtensor<std::int32_t, 2>* puppet_facets;
+    auto element = _V->element();
+    const std::uint32_t bs = element->block_size();
+    mesh->topology_mutable().create_entity_permutations();
+
+    auto cell_type
+        = dolfinx::mesh::cell_type_to_basix_type(mesh->topology().cell_type());
+    // Get facet normals on reference cell
+    auto facet_normals = basix::cell::facet_outward_normals(cell_type);
+
+    // Select which side of the contact interface to loop from and get the
+    // correct map
+    if (origin_meshtag == 0)
+    {
+      map = &_map_0_to_1;
+      q_phys_pt = &_qp_phys_0;
+      puppet_facets = &_cell_facet_pairs_0;
+    }
+    else
+    {
+      map = &_map_1_to_0;
+      q_phys_pt = &_qp_phys_1;
+      puppet_facets = &_cell_facet_pairs_1;
+    }
+    std::size_t max_links = std::max(_max_links_0, _max_links_1);
+    const std::int32_t num_facets = (*map).shape(0);
+    const std::int32_t num_q_points = _qp_ref_facet[0].shape(0);
+    const std::int32_t ndofs = _V->dofmap()->cell_dofs(0).size();
+    std::vector<PetscScalar> c(num_facets * num_q_points * gdim, 0.0);
+    const int cstride = num_q_points * gdim;
+    std::vector<std::int32_t> perm(num_q_points);
+    std::vector<std::int32_t> sorted_cells(num_q_points);
+    xt::xtensor<double, 2> q_points
+        = xt::zeros<double>({std::size_t(num_q_points), std::size_t(gdim)});
+    xt::xtensor<double, 2> dphi;
+    xt::xtensor<double, 3> J = xt::zeros<double>(
+        {std::size_t(1), std::size_t(gdim), std::size_t(tdim)});
+    xt::xtensor<double, 3> K = xt::zeros<double>(
+        {std::size_t(1), std::size_t(tdim), std::size_t(gdim)});
+    xt::xtensor<double, 1> detJ = xt::zeros<double>({1});
+    xt::xtensor<double, 4> phi(cmap.tabulate_shape(1, 1));
+    for (int i = 0; i < num_facets; i++)
+    {
+      auto cell = (*puppet_facets)(i, 0); // extract cell
+
+      for (int j = 0; j < num_q_points; j++)
+      {
+        sorted_cells[j] = (*map)(i, j, 0);
+        // Is there a better way to do this???
+        for (int k = 0; k < gdim; k++)
```

Comment:
Is "this" Computing the the quadrature points on the other facet (i.e. summing quadrature point for a given facet with the gap)?

I think this function would be cleaner if it was not part of the contact class, and simply took in the cells and corresponding points on each cell, and outputted the normals.

Reply:
This "standalone" function could start around line 1360 i think

Reply:
This will be addressed separately. See #26.